### ISSUE TITLE
Specify per dataset which field choices can be selected for the each of the phonology fields

### ISSUE THREAD

**Issue by ocrasborn** [Signbank/Global-signbank#340]
In Django, but ideally in a separate page on the site where the dataset manager (non-existing group/role right now) can specify these (and other things, like which spoken language is used on the site in addition to English to describe signs, and which interface languages are selectable).

**Comment by ocrasborn** [comment]
Perhaps in a table with datasets in columns and field values in rows, where checkboxes can be ticked for each value that is available for a given dataset?
The goal is to not bother users with, say, location values that do not occur in a specific language during editing a sign. (What to make available in search forms is a separate matter, we'll need to discuss that when revising the search functions.)

**Comment by ocrasborn** [comment]
Created a group Dataset Manager today which can be assigned permissions to select field choices when this issue has been addressed.

**Comment by vanlummelhuizen** [comment]
I have given this issue some thought. 

First a question: is it clear from the models what the phonology fields are? Or to put it a bit differently: should we make a special distinction between phonology fields and other fields?

I have also taken in consideration what @henrinie said in https://github.com/Signbank/Global-signbank/issues/381#issuecomment-334146283. It may be worthwhile to overhaul the FieldChoice mechanics, e.g. in the following way: 

1. Take the *field* of a FieldChoice out of FieldChoice and turn it into a FieldChoiceType model. No need to retype the *field* for every FieldChoice (which is error prone).
2. Use https://docs.djangoproject.com/en/1.11/ref/forms/fields/#modelchoicefield for FieldChoice. As I understand it, it determines a list of choices based on a queryset. It would be no longer necessary to restart Signbank after adding/changing field choices. Moreover, it would help with the next point.
3. For this issue, it is necessary to keep track of the field choices that are to be used with a specific dataset. We could make a Dataset-FieldChoice relation model to implement this. This could also be implemented as muting (opt-out) field choices for a specific dataset. In that case, by default everything is available.
4. Use model-translation for the field choice names. Now, English, Dutch and Chinese are still hard coded.

Any comments @Woseseltops, @ocrasborn ?

**Comment by Woseseltops** [comment]
You have my approval @vanlummelhuizen . Keep in mind though that overhauls tend to lead to nightmares for the ASL Signbank, so you might want to implement these changes after #400 . 

> This could also be implemented as muting (opt-out) field choices for a specific dataset.

Yes I thinking the same way. So you can give for each dataset a list of fieldchoices that are not relevant.

**Comment by ocrasborn** [comment]
Some responses to the points of @vanlummelhuizen : 
- It is not clear by the field name in the model what the phonology fields are, if that's what you're asking. Or is the model internally structured somewhere that I haven't seen before? (This reminds me of something I had planned to do long ago: make an overview of the differences between the different Signbanks, esp. in terms of phonology. Haven't done so yet. Will try to do so this week.)
- Re: 1: that would be nice.
- Re: 3: opt-out is fine, defaulting to having all values for each dataset.

**Comment by ocrasborn** [comment]
Added requirement: we should be able to limit access to this to a certain group of users (Dataset_manager in fact), in such a way that the dataset manager can specify which values are not available for the datasets that s/he also has write access to.

**Comment by Woseseltops** [comment]
Okay this has been a high priority issue for almost a year now, but I think nobody (incl myself) picks it up because there are no concrete actions. Let's write down some concrete actions:
1. Add a ManyToMany field 'exclude_choices' to the Dataset model. Initially, you can fill these field from the admin with field choices.
2. During the generation of choice lists on the detail view, it takes into account the dataset of the current gloss, and exludes the choices in the 'exclude_choices' field for that dataset.
3. In the dataset manager page, you can select the content for 'exclude_choices' with some multivalue input field.

Some other things mentioned here I think deserve their own issue (agree @vanlummelhuizen ?):
* Make FieldChoice field `field` a seperate model instead of a string.
* Use model-translation for FieldChoice field `name`

**Comment by Woseseltops** [comment]
Step 1 is done: you can no exclude field choices from a dataset in the admin; requires a migration @susanodd and @vanlummelhuizen ! By the way, @susanodd , I uncommented some code you had commented out in bb7f2770 ; I hope I didn't break anything but all tests are still green.

Next step is actually hiding the excluded field choices in the detail view. I will only do this in the seperate field choice page I'm creating for #504 , but this means this page will have to have an extra dataset parameter.

**Comment by ocrasborn** [comment]
Thanks @Woseseltops for taking this up. A proposal for an additional action:
4. Create an interface for Dataset Managers to show/hide field values in the interface, without needing access to the admin. (Proposal: a table view with values in rows and datasets in columns, so that people can "manage" this for all the datasets they are manager for.)

**Comment by Woseseltops** [comment]
By the way, step 2 is now also done: field choices set to hidden for a dataset no longer show up for glosses in that dataset.

**Comment by Woseseltops** [comment]
https://signbank.science.ru.nl/datasets/field_choices/ I've made quite some progress with step 3-4 . The table already represents what's actually in the database... it's just that nothing happens when you click 'save' :)

**Comment by Woseseltops** [comment]
Okay now the save button really does something. 

Next up: 
* make the code more safe in case something does NOT go as planned
* check: only your own datasets?
* Put the link on a logical place in the interface.

**Comment by Woseseltops** [comment]
Check check check... done?

Two things I'm hesitating about, curious for your opinion @ocrasborn : 
* I've added this page as a button to the dataset manager page... is this the best place?
* Because there are so many field choices, this table approach is not convenient. Some kind of filtering functionality would help... but would this functionality be used enough to justify these efforts (which are not trivial?

**Comment by ocrasborn** [comment]
Hi @Woseseltops : thanks! This is another step forward in making the Signbank more Global. In response to your questions:
* Yes, excellent location
* Hm, yes, I agree that this is a dilemma. For now, let's try to do something simple (I think): freeze the header (of Signbank in general, incl. the menu bar, and of the table), and let the table scroll under it. That will help for people with more than one dataset to manage. Indeed the effort won't be undertaken so often, so unless a ship with money comes in, let's postpone a nicer way to filter the table. Could you create a new issue for it, though?

**Comment by Woseseltops** [comment]
Wow, freezing that header was a lot harder than I imagined, but I've found a good and simple solution now. Issue is also created. Happy @ocrasborn ?

**Comment by ocrasborn** [comment]
Almost happy, @Woseseltops:
* The scrolling pane occupies half the page on my large screen, with the site footer correctly appearing at the bottom. Is it possible to make the height of the scrolling pane dynamic?
* Could you block the page for anonymous users? It now displays (if you know the url) the list of values without any columns with checkmarks, as anonymous users cannot have a dataset selected

**Comment by Woseseltops** [comment]
> The scrolling pane occupies half the page on my large screen, with the site footer correctly appearing at the bottom. Is it possible to make the height of the scrolling pane dynamic?

Done

> Could you block the page for anonymous users? It now displays (if you know the url) the list of values without any columns with checkmarks, as anonymous users cannot have a dataset selected

Done

Happy @ocrasborn ?

**Comment by Woseseltops** [comment]
We just saw on @ocrasborn's screen that the header doesn't 'stick' in Safari. Just verified that it does in both Chrome and Firefox. I'm leaning towards leaving it like this, because this page is barely used anyway. Agreed?

**Comment by ocrasborn** [comment]
Agree @Woseseltops
